File: scala/palindrome-products/src/main/scala/PalindromeProducts.scala
case class PalindromeProducts(lower: Int, upper: Int) {
  private val factors =
    for {
      (x, i1) <- (lower to upper).zipWithIndex
      (y, i2) <- (lower to upper).zipWithIndex
      if i1 <= i2
    } yield (x, y)

  private val palindromes =
    factors
      .filter { case (x, y) => isPalindrome(x * y) }
      .groupBy { case (x, y) => x * y }
      .map { case (product, factors) => (product, factors.toSet) }
      .toSeq
      .sortBy(_._1)

  def smallest: Option[(Int, Set[(Int, Int)])] = palindromes.headOption

  def largest: Option[(Int, Set[(Int, Int)])] = palindromes.lastOption

  private def isPalindrome(i: Int): Boolean = {
    val digits = i.toString
    digits.reverse == digits
  }
}

File: fsharp/palindrome-products/PalindromeProducts.fs
module PalindromeProducts

let private reverse =
    Array.ofSeq >> Array.rev >> System.String

let private isPalindrome n =
    let digits = sprintf "%d" n
    reverse digits = digits

let private products min max =
    if min > max then
        invalidArg "min" "invalid range"
    else
        [ for x in min .. max do
            for y in min .. max do
                if x <= y && isPalindrome (x * y) then yield x, y ]

let private palindromes min max =
    products min max
    |> List.groupBy (fun (x, y) -> x * y)
    |> List.sortBy fst

let largest min max =
    match List.tryLast (palindromes min max) with
    | Some (product, factors) -> Some product, factors
    | None -> None, []

let smallest min max =
    match List.tryHead (palindromes min max) with
    | Some (product, factors) -> Some product, factors
    | None -> None, []